### PR TITLE
Parse NaN from object like {"n": NaN}

### LINKED ISSUE #24: Parse "NaN"
Python's JSON parser and others understand:
```
{"x": NaN}
```
I can offer a pull request for JSONSL to understand it too, if you're interested.

### PR DESCRIPTION
**Pull request by ajdavis**
Parsing is case-insensitive. Closes #24.

### CODE REVIEW

**Comment by ajdavis** [comment]
This is ready for an initial review. "make check" passes, but there's not yet any test for the new NaN parsing. The next steps are to control this NaN parsing with a compile-time or runtime option, and to add a test for it that expects success or failure depending on whether NaN parsing is enabled. 

Would you like to do those steps, or tell me how I should do them?

File: jsonsl.c [hunk 1]
```diff
@@ -424,6 +424,21 @@ jsonsl_feed(jsonsl_t jsn, const jsonsl_char_t *bytes, size_t nbytes)
...
+                      INVOKE_ERROR (SPECIAL_EXPECTED);
+                   }
+                } else if (state->special_flags & JSONSL_SPECIALf_NULL ||
+                           state->special_flags & JSONSL_SPECIALf_NAN) {
+                   /* previous char was "n", are we parsing null or nan? */
+                   if (CUR_CHAR != 'u') {
+                      state->special_flags &= ~JSONSL_SPECIALf_NULL;
+                   }
+
+                   if (tolower(CUR_CHAR) != 'a') {
```
Comment (mnunberg) [inline jsonsl.c]: `else if`

File: jsonsl.c [hunk 2]
```diff
@@ -1425,11 +1440,13 @@ static unsigned short Special_Table[0x100] = {
         /* 0x37 */ JSONSL_SPECIALf_UNSIGNED /* <7> */, /* 0x37 */
         /* 0x38 */ JSONSL_SPECIALf_UNSIGNED /* <8> */, /* 0x38 */
         /* 0x39 */ JSONSL_SPECIALf_UNSIGNED /* <9> */, /* 0x39 */
-        /* 0x3a */ 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, /* 0x59 */
+        /* 0x3a */ 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+        /* 0x4e */ JSONSL_SPECIALf_NAN /* <N> */, /* 0x4e */
```
Comment (mnunberg) [inline jsonsl.c]: FYI this table is generated by a Perl script. Better to modify it there.

File: jsonsl.c [hunk 3]
```diff
@@ -1425,11 +1440,13 @@ static unsigned short Special_Table[0x100] = {
...
+        /* 0x4f */ 0,0,0,0,0,0,0,0,0,0,0, /* 0x59 */
         /* 0x5a */ 0,0,0,0,0,0,0,0,0,0,0,0, /* 0x65 */
         /* 0x66 */ JSONSL_SPECIALf_FALSE /* <f> */, /* 0x66 */
         /* 0x67 */ 0,0,0,0,0,0,0, /* 0x6d */
-        /* 0x6e */ JSONSL_SPECIALf_NULL /* <n> */, /* 0x6e */
+        /* 0x6e */ JSONSL_SPECIALf_NULL | JSONSL_SPECIALf_NAN /* <n> */, /* 0x6e */
```
Comment (mnunberg) [inline jsonsl.c]: likewise

**Comment by mnunberg** [comment]
`JSONSL_PARSE_NAN` would be the compiler directive.

Regarding tests - all existing parse tests load from a file. You'd make a new file and simply feed a `[nan]` or similar, verifying it parses successfully. This reminds me that I should add better tests for type-specific valdiation (right now all our tests do is ensure that things parse successfully (or not)).

**Comment by mnunberg** [comment]
oops. didn't mean to close

File: jsonsl.c (hunk 1 above)
Comment (ajdavis) [inline jsonsl.c]: Actually it's two "!=" statements, so CUR_CHAR might be neither 'a', 'N', nor 'n', I think we have to check both if statements.

**Comment by ajdavis** [comment]
I see the tests didn't work on Travis as they did on my machine, investigating....

**Comment by ajdavis** [comment]
OK, this pull request is ready for review.

**Comment by mnunberg** [comment]
Can you squash these into a single commit?

Also, I'm still a bit uneasy about having the NaN parsing around there as dead code even if it's not being compiled that way. I have another idea. How about making a proxy variable.. e.g.

(in the character table for `special_begin`):

`JSONSL_SPECIALf_NULL|JSONSL__NAN_PROXY`

Then..

```
#ifdef JSONSL_PARSE_NAN
#define JSONSL__NAN_PROXY JSONSL_SPECIALf_NAN
#else
#define JSONSL__NAN_PROXY 0
#endif
```

This way, the current (pre-nan) code can remain the same, and newer code/branches can be added on only if `PARSE_NAN` is defined. The way you have it currently, it will run through some cycles determining NAN stuff even if nan parsing is disabled. My desire to have this be a compile time option is less about one of configuration and more about not impacting performance (even theoretically) for less common situations (granted, `null` isn't super common either..).

So basically:
```
#ifdef JSONSL_PARSE_NAN
else if (state->special_flags & JSONSL_SPECIALf_NAN) {
    // ...
}
#endif
```

Because `_NAN_PROXY` is 0 if `PARSE_NAN` is not defined, the character table will only yield `JSONSL_SPECIALf_NULL` on `n` and nothing on `N`.

**Comment by ajdavis** [comment]
Done, let me know if I followed your instructions correctly.

**Comment by mnunberg** [comment]
yep! exactly what I had in mind. thanks for the contribution!
